File: 988. Smallest String Starting From Leaf/smallestFromLeaf.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
    res = '~'
    def dfs(node, path):
      nonlocal res
      if not node:
        return
      
      path.append(chr(97 + node.val))

      if not node.left and not node.right:
        pathStr = ''.join(reversed(path))
        res = min(res, pathStr)
        
      if node.left:
        dfs(node.left, path)
        
      if node.right:
        dfs(node.right, path)
        
      path.pop()
    
    dfs(root, [])
    return res
  

root = TreeNode(0)
root.left = TreeNode(1)
root.right = TreeNode(2)

root.left.left = TreeNode(3)
root.left.right = TreeNode(4)

root.right.left = TreeNode(3)
root.right.right = TreeNode(4)

solution = Solution()
print('smallestFromLeaf: ', solution.smallestFromLeaf(root))

File: 988. Smallest String Starting From Leaf/smallestFromLeaf.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {string}
 */
const smallestFromLeaf = (root) => {
  let res = '~';
  const dfs = (node, path) => {
    if (!node) {
      return;
    }

    path.push(String.fromCharCode(97 + node.val));

    if (!node.left && !node.right) {
      const pathCopy = [...path];
      const pathStr = pathCopy.reverse().join('');
      res = pathStr < res ? pathStr : res;
    }

    if (node.left) {
      dfs(node.left, path);
    }

    if (node.right) {
      dfs(node.right, path);
    }

    path.pop();
  };

  dfs(root, []);

  return res;
};

const root = new TreeNode(0);
root.left = new TreeNode(1);
root.right = new TreeNode(2);

root.left.left = new TreeNode(3);
root.left.right = new TreeNode(4);

root.right.left = new TreeNode(3);
root.right.right = new TreeNode(4);

console.log('smallestFromLeaf: ', smallestFromLeaf(root));
